### PR TITLE
net-dns/bind: bump to 9.18.29

### PR DESCRIPTION
**Pull request by thesamesam**
Supersedes #24001.

### CODE REVIEW

**Comment by thesamesam** [comment]
still need to test and fixup init scripts for chroot, I think

**Comment by WhitleyStriber** [comment]
Hi @thesamesam ,

This is metafest from yesterday in IRC. There are a few bugs I found with the ebuild you helped me with for 9.18.9:

**1. When building with the `dnsrpz` flag, BIND fails to build with this error:**

error while building with dnsrps: configure: error: dlopen and librpz.so needed for DNSRPS

checking whether to enable query trace logging... no
checking for librpz __attribute__s... yes
checking for library containing librpz_client_create... no
configure: error: dlopen and librpz.so needed for DNSRPS

!!! Please attach the following file when seeking support:
!!! /var/tmp/portage/net-dns/bind-9.18.9/work/bind-9.18.9/config.log
 * ERROR: net-dns/bind-9.18.9::local failed (configure phase):
 *   econf failed
 *
 * Call stack:
 *               ebuild.sh, line  136:  Called src_configure
 *             environment, line 1361:  Called econf '--prefix=/usr' '--sysconfdir=/etc/bind' '--localstatedir=/var' '--enable-full-report' '--without-readline' '--with-openssl=/usr' '--with-json-c' '--with-zlib' '--enable-linux-caps' '--enable-dnsrps' '--enable-dnstap' '--disable-doh' '--without-libnghttp2' '--enable-fixed-rrset' '--disable-static' '--enable-geoip' '--with-maxminddb' '--with-gssapi' '--without-libidn2' '--without-jemalloc' '--without-lmdb' '--without-libxml2'
 *        phase-helpers.sh, line  730:  Called __helpers_die 'econf failed'
 *   isolated-functions.sh, line  112:  Called die
 * The specific snippet of code:
 *              die "$@"

**2. When you install BIND-9.18.9 and try to start it, it fails finding 2 files: /etc/bind/named.conf and /var/bind/named.cache with this log:** 

17-Dec-2022 07:43:48.339 general: warning: TKEY mode 2 support (Diffie-Hellman): yes
17-Dec-2022 07:43:48.339 general: warning: TKEY mode 3 support (GSS-API): no
17-Dec-2022 07:43:48.339 config: warning: config.c: option 'trust-anchor-telemetry' is experimental and subject to change in the future
17-Dec-2022 07:43:48.339 general: info: loading configuration from '/etc/bind/named.conf'
17-Dec-2022 07:43:48.339 config: error: open: /etc/bind/named.conf: file not found
17-Dec-2022 07:43:48.339 config: error: open: /var/bind/named.cache: file not found
17-Dec-2022 07:43:48.339 general: critical: loading configuration: file not found
17-Dec-2022 07:43:48.339 general: critical: exiting (due to fatal error)

could not configure root hints from var vind named.cache file not found

I have not tested chroot so far, but I've been running BIND for about 2 hours now with no problems on my home network. If I find any issues I'll post here.

**Comment by WhitleyStriber** [comment]
Sam, it looks like a lot of config files now are in /var/bind instead of /etc/bind. When i create a log or try to include a zonefile, i tend to get errors saying permission denied or file not found, because it's trying to find them in /var/bind. hopefully this helps.

**Comment by thesamesam** [comment]
Thanks! I'll take a look at that over the next few days hopefully.

**Comment by WhitleyStriber** [comment]
The var/bind issue is a normal setup. I have not tested chroot because like i said above, the config file does not have that CHROOT variable line as portage outputs:

 * You can edit /etc/conf.d/named to customize named settings
 * 
 * If you'd like to run bind in a chroot AND this is a new
 * install OR your bind doesn't already run in a chroot:
 * 1) Uncomment and set the CHROOT variable in /etc/conf.d/named.
 * 2) Run `emerge --config '=net-dns/bind-9.18.9'`

Here is the file that I have for /etc/conf.d/named:

# Set various named options here.
#
#OPTIONS=""

# config file path
NAMED_CONF="/etc/bind/named.conf"

# Set this to the number of processors you want bind to use.
# Leave this unchanged if you want bind to automatically detect the number
#CPU="1"

# Default pid file location
# use named.conf to specify pid-file location

# Scheduling priority: 19 is the lowest and -20 is the highest.
# Default: 0
#NAMED_NICELEVEL="0"

**Comment by thesamesam** [comment]
Thanks, I don't see where you mentioned the CHROOT line earlier (just that you hadn't tried it yet), but no worries. I'll check that out too.

**Comment by WhitleyStriber** [comment]
I had issues with bind not starting correctly, so I wiped out /etc/bind and /var/bind folders and emerged again. When I start, I get this:

EmmaHarben /etc/bind # rc-service named start
 * Caching service dependencies ...                                                                                                                                                           [ ok ]
 * Starting named ...
 * Checking named configuration ...
 * No /etc/bind/named.conf file exists!                                                                                                                                                       [ !! ]
 * ERROR: named failed to start

Basically I think 90% of the problem is just putting config files in the right places. Thanks sam, I'm gonna keep testing.

**Comment by WhitleyStriber** [comment]
> Thanks, I don't see where you mentioned the CHROOT line earlier (just that you hadn't tried it yet), but no worries. I'll check that out too.

you're right, i forgot to post my chroot bug -- stupid me

**Comment by Delicates** [comment]
Just a sanity check:
Does CHROOT still need /dev/urandom creation?
See:  https://gitlab.isc.org/isc-projects/bind9/-/issues/3399

Or is this still used on some platforms or old kernels that need it?
If so, why aren't we creating both /dev/random and /dev/urandom in CHROOT?

Or are we keeping it because once CHROOT is created by old ebuilds we can't remove /dev/urandom from it without breakage?

Tried deleting /dev/urandom to test if named would work without it, but the init script gave me an error:

* Your chroot dir /chroot/dns is inconsistent, please run 'emerge --config net-dns/bind' first

I checked lsof on a running named - it had only /dev/null opened, but neither /dev/urandom nor /dev/zero.

**Comment by Delicates** [comment]
Also now that rndc-confgen no longer has the -r option, is this notice still valid, or should it be rewritten?

```einfo "Using /dev/urandom for generating rndc.key"```

See: https://gitlab.isc.org/isc-projects/bind9/-/merge_requests/269/diffs?commit_id=3a4f820d625c214cfb21f5e6d18ce9160d2a193b#diff-content-09ca95f2dbb6ddcbf4ade21842384fd45d708a9b

**Comment by Delicates** [comment]
Found the following write-up within the BIND distro file: doc/arm/security.inc.rst
Interestingly it talks about /dev/random, not about /dev/urandom (not sure if it is outdated):

```
Unlike with earlier versions of BIND, 
:iscman:`named` does *not* typically need to be compiled statically, nor do shared libraries need to be installed under the new
root. However, depending on the operating system, it may be necessary to set
up locations such as ``/dev/zero``, ``/dev/random``, ``/dev/log``, and
``/etc/localtime``.
```

grepping all source code files for these /dev/ nodes doesn't really come up with anything (unless they are used by linked dependencies).

**Comment by thesamesam** [comment]
OK, I've cleaned up the random stuff a bit. Not had a chance to look at config file location issues yet.

**Comment by WhitleyStriber** [comment]
sam if you have a new package for me to test with all the new config files for CHROOT i'm more than happy to test.

**Comment by thesamesam** [comment]
> sam if you have a new package for me to test with all the new config files for CHROOT i'm more than happy to test.

I've put a new tarball at https://dev.gentoo.org/~sam/tmp/bind.tar.xz, thanks!

**Comment by Delicates** [comment]
> I've put a new tarball at https://dev.gentoo.org/~sam/tmp/bind.tar.xz, thanks!

Would be happy to test upgrading, but can't until the bind-tools issue is resolved. I have things installed that depend on bind-tools.

**Comment by Delicates** [comment]
Thanks sam!

Good thing that prior to trying I took a backup of my:
* `/var/bind/`
* `/etc/bind/`
* `/etc/conf.d/named`

It saved my bacon!
I noted down 5 things worth mentioning.

After some fiddling I got it to work.
1. Prior to build bumped both ebuilds to version 9.18.12
2. Prior to build enabled `doh` USE flag for bind (thanks for adding this USE flag!) - built fine, haven't tested functionality.

Trying to emerge bind:
image

Emerging bind-tools:
image

Everything emerged fine, however had to fix 2 things to be able to run it:
2. The old path of `/usr/sbin/named-checkconf` now seems to be `/usr/bin/named-checkconf` so the init and the service scripts need to be updated.
3. The installation caused the following data loss (had to restore these 3 files from backup):
```
<<<          sym /var/bind/root.cache
<<<          obj /var/bind/pri/localhost.zone
<<<          obj /var/bind/named.cache
```

Additionally:
4. While troubleshooting, I have also noticed log error messages pointing to a CHROOT permissions problem (exists since prior to 9.18):
```
named[3287]: Could not open '/var/run/named/session.key'.
named[3287]: Please check file and directory permissions or reconfigure the filename.
named[3287]: could not open file '/var/run/named/session.key': Permission denied
named[3287]: could not create /var/run/named/session.key
named[3287]: failed to generate session key for dynamic DNS: permission denied
```
After changing ownership from `root` to `named` for the following two paths, there is no more errors in the log:
* `${CHROOT}/var/run/named/`
* `${CHROOT}/var/run/named/session.key`

Possible ebuild enhancement for the future (not critical for now):
5. As part of named startup, `named-checkconf` has started to show the following alerts:
```
/etc/bind/named.conf:140: option 'auto-dnssec' is deprecated
/etc/bind/named.conf:140: 'auto-dnssec' option is deprecated and will be removed in BIND 9.19. Please migrate to dnssec-policy
```

It may be good to consider including running or the freshly built `named-checkconf` as part of post-build/pre-install checks IF the build is an upgrade of a previously installed BIND version:
* If return status is 0 - print any output as build warnings.
* If return status is 1 - abort installation and print out the error output to avoid the outage.

If `/etc/bind/named.conf` doesn't exist (e.g. fresh BIND build) - no need to run it.

**Comment by gjaekel** [comment]
I set up this files for bind-9.18.20 and it build to a package, replace a bind-9.16.42 with it. And it seems to work proper (at least 10mins :smile:).

As written above, there is a small bug in `/etc.init.d/named`: One have to replace all occurrences of  `/usr/sbin/named-checkconf` by `/usr/bin/named-checkconf`.

**Comment by Delicates** [comment]
> Everything emerged fine, however had to fix 2 things to be able to run it:
> 3. The old path of `/usr/sbin/named-checkconf` now seems to be `/usr/bin/named-checkconf` so the init and the service scripts need to be updated.

This wasn't addressed. Init script are currently broken in portage.

**Comment by thesamesam** [comment]
Fixed.

**Comment by gjaekel** [comment]
@thesamesam: Thank you!
